From: Casey Bodley <cbodley@redhat.com>
Date: Tue, 11 Oct 2016 15:19:37 -0400
Subject: rgw: clear data_sync_cr if RGWDataSyncControlCR fails

async notifications will still try to call wakeup() on RGWDataSyncControlCR
if it fails, leading to segfault

Fixes: http://tracker.ceph.com/issues/17569

Signed-off-by: Casey Bodley <cbodley@redhat.com>
(cherry picked from commit 5cc599b9bf2dde31de16a5b2831baf06851d69d5)

Resolves: rhbz#1393665
---
 src/rgw/rgw_data_sync.cc | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/src/rgw/rgw_data_sync.cc b/src/rgw/rgw_data_sync.cc
index 101ddb0..96a8c1b 100644
--- a/src/rgw/rgw_data_sync.cc
+++ b/src/rgw/rgw_data_sync.cc
@@ -1494,16 +1494,17 @@ int RGWRemoteDataLog::run_sync(int num_shards, rgw_data_sync_status& sync_status
   lock.get_write();
   data_sync_cr = new RGWDataSyncControlCR(&sync_env, num_shards);
   lock.unlock();
+
   r = run(data_sync_cr);
-  if (r < 0) {
-    ldout(store->ctx(), 0) << "ERROR: failed to run sync" << dendl;
-    return r;
-  }
 
   lock.get_write();
   data_sync_cr = NULL;
   lock.unlock();
 
+  if (r < 0) {
+    ldout(store->ctx(), 0) << "ERROR: failed to run sync" << dendl;
+    return r;
+  }
   return 0;
 }
 
